Add root parameter to useradd, shadow and groupadd

aplanas opened the pull request:
### What does this PR do?
Add root parameter to useradd and shadow execution modules

aplanas committed:
shadow: add root parameter to the public API

Refactor the code to avoid duplications, and use the alternative
spwd functions when a different root is set.

cachedout commented on salt/modules/shadow.py:
```diff
@@ -96,15 +96,19 @@ def set_inactdays(name, inactdays):
-    cmd = 'chage -I {0} {1}'.format(inactdays, name)
+    cmd = ['chage']
+    if root is not None and __grains__['kernel'] != 'AIX':
+        cmd.extend(('-R', root))
```
I cannot find documentation for `-R` when used with `chage`. Could you please point me to some?

aplanas replied:
Sure!

I can see it in my system (openSUSE) and in the source code:

https://github.com/shadow-maint/shadow/blob/master/src/chage.c#L159

According to blame was added 7 years ago:

https://github.com/shadow-maint/shadow/commit/0857837e64a5a45ef686806aa3f3694ff733d5b2

cachedout replied:
Hrm. My worry here is that I'm seeing this in certain docs:

https://linux.die.net/man/1/chage

That said, perhaps we could just improve error checking down in L104 and we should be fine. Thoughts?

aplanas replied:
I do not know what to say, looks like that was documented 7 years ago too, according to the previous commit.

I can see the same option in CentOS 7, that provides shadow-utils-4.1.5.1-24.

I do not know from where https://linux.die.net gets the information, or when was the last update, but I cannot find the same option in useradd, another tool form shadow-utils:

https://linux.die.net/man/8/useradd

Also note that there is a missing CONFIGURATION section at the end of the man page, that was added 9 years ago:

https://github.com/shadow-maint/shadow/commit/5d6c3143047c30c2c402405a98806bdd664b6bee

But more important, the same -R parameter is already in the salt code for useradd (that is also part of the shadow-utils package), added three years ago:

https://github.com/saltstack/salt/commit/1eb3b9c5e4ae3952c8198692e994997207733406

IMHO linux.dice.net is not a reliable source of information.

P.S according to https://www.die.net/ the man pages are from 2002

aplanas replied:
> That said, perhaps we could just improve error checking down in L104 and we should be fine. Thoughts?

This makes me wonder what to do with the already -R in useradd, that was added the same date in a different commit:

https://github.com/shadow-maint/shadow/commit/cecae46ccfb85c2c4bf6e0fe0419d740ba1f2ecc

aplanas commented on salt/modules/useradd.py:
```diff
@@ -470,10 +473,10 @@ def chfullname(name, fullname):
         salt '*' user.chfullname foo "Foo Bar"
     '''
-    return _update_gecos(name, 'fullname', fullname)
+    return _update_gecos(name, 'fullname', fullname, root)
```
Uhmm this is not correct. I realized that _update_gecos accept the root parameter, used to usermod, but this is deceptive as the return value use _get_gecos, that use pwd.getpwnam, that do not have a root parameter.

Maybe I can implement an alternative to pwd.* in case that root parameter is provided, and if not falback to pwd

aplanas committed:
useradd: add alternative versions for pwd functions

aplanas committed:
useradd: normalize the use of root parameter

Use the root parameter in the rest of the public API.

Refactor the code to remove duplicate code.

aplanas committed:
shadow: add alternative versions for spwd functions

aplanas committed:
groupadd: add alternative versions for grp functions

aplanas committed:
groupadd: normalize the use of root parameter

aplanas committed:
groupadd: user --root instead of -R or -Q

In some places of the code gpasswd and usermod are used depending
on the plataform.  One use -Q and other use -R to indicate a
different root parameter.

This patch unify this, using --root parameter.

Partially fix #43131

aplanas committed:
shadow: fix name missmatch in Python3

In Python 3 the fields for name and password for the spwd module
are different that in Python 2.

This patch makes sure to explore first the Python 2 name and later
check the Python 3 if is not found.

aplanas commented:
Ready for review.

cachedout commented:
@aplanas Could you add docstrings for these new parameters, please?

aplanas commented:
@cachedout I added the documentation strings for root and for the missing parameters. Done.

aplanas commented:
Checking the logs looks like that user.getent fails in windows, but this PR do not touch this code (is in a different module)

cachedout commented:
@aplanas I've asked @twangboy from @saltstack/team-windows to swing by here and see if he can sort out why that test is failing before we get this merged.

aplanas commented:
any more reviews? Any thing that I can do?

twangboy commented:
I don't think this will cause any problems in Windows

isbm commented on salt/modules/groupadd.py:
```diff
@@ -123,41 +165,74 @@ def getent(refresh=False):
     ret = []
-    for grinfo in grp.getgrall():
+    if root is not None:
+        _fun = functools.partial(_getgrall, root=root)
```
Even GvR himself [doesn't agree](https://twitter.com/gvanrossum/status/391769557758521345?lang=en) with you here! :laughing: Me too, BTW.

aplanas replied:
Can I disagree here? The example is for partials that do not expect arguments (is done explicit two or three comments later), and this is not the case. I find this more explicit than this:

```python
    if root is not None:
        _fun = lambda name: _getgrnam(name, root=root)
    else:
        _fun = lambda name: grp.getgrnam(name)
```

In the lambda form is not explicit that the call is not evaluated (lazy evaluation), and that the free parameter is indeed name.

If you have an strong objection I can change it.

```

aplanas replied:
Umh. E731 (do not assign a lambda expression, use a def) happens in this case.

I like `lambda`s as anyone, but in this case if some new user see the code, she/he will find documentation about [functools.partial](https://docs.python.org/3.7/library/functools.html#functools.partial) easily.

Tangentially about this topic, is there any set of WXXX or EXXX that the team agrees to ignore?

isbm replied:
Ahh, you're [running into this](https://www.python.org/dev/peps/pep-0008/#programming-recommendations).

```python
if root:
    def _fun(name): return _getgrnam(name, root=root)
else:
    def _fun(name): return grp.getgrnam(name)
```

aplanas replied:
:confused:
What how is defining a local inner function be better than a standard library that is documented? Is there any plan to drop `functools.partial` that I am not aware of?

Also is used along the codebase of salt too:

* https://github.com/saltstack/salt/blob/develop/salt/utils/parsers.py#L24
* https://github.com/saltstack/salt/blob/develop/salt/utils/boto3mod.py#L40
* https://github.com/saltstack/salt/blob/develop/salt/utils/botomod.py#L40
* https://github.com/saltstack/salt/blob/develop/salt/utils/roster_matcher.py#L81
* https://github.com/saltstack/salt/blob/develop/salt/modules/rpmbuild.py#L557
...

isbm replied:
@aplanas I am just fixing you a lint error according to as [PEP-8 says](https://www.python.org/dev/peps/pep-0008/#programming-recommendations). I still find this:

```python
def foo(x): f(x)
```

Much better than this:

```python
foo = functools.partial(f, x)
```

aplanas replied:
For future reference. I found the pep about currying:

https://www.python.org/dev/peps/pep-0309/

isbm commented on salt/modules/groupadd.py:
```diff
@@ -123,41 +165,74 @@ def getent(refresh=False):
+
+    cmd.extend((param, value, name))
+
+    __salt__['cmd.run'](cmd, python_shell=False)
```
Some checks/logging would help here for debugging.

aplanas replied:
Indeed, can we do that in a different change? The original code do not provide any check or log here for useradd, groupadd and shadow. I this this change for avoid code duplication, but I take care or not changing the semantic.

Adding some checks here can be normalized in a different commit, and this will require adjusting the tests.

isbm commented on salt/modules/groupadd.py:
```diff
@@ -284,3 +377,37 @@ def members(name, members_list, root=None):
+    '''
+    Alternative implementation for getgrnam, that use only /etc/group
+    '''
+    root = '/' if not root else root
```
Safe to just say:

```python
root = root or '/'
```

aplanas replied:
sure

isbm commented on salt/modules/groupadd.py:
```diff
@@ -284,3 +377,37 @@ def members(name, members_list, root=None):
+                comps[2] = int(comps[2])
+                comps[3] = comps[3].split(',') if comps[3] else []
+                return grp.struct_group(comps)
+    raise KeyError
```
Please add message here, so it gets logged later.

aplanas replied:
yes

isbm commented on salt/modules/groupadd.py:
```diff
@@ -284,3 +377,37 @@ def members(name, members_list, root=None):
+    '''
+    Alternative implemetantion for getgrall, that use only /etc/group
+    '''
+    root = '/' if not root else root
```
Also can be just `or`.

isbm commented on salt/modules/groupadd.py:
```diff
@@ -284,3 +377,37 @@ def members(name, members_list, root=None):
+            comps = line.strip().split(':')
+            # Generate a getgrall compatible output
+            comps[2] = int(comps[2])
+            comps[3] = comps[3].split(',') if comps[3] else []
```
Scary. Here I would first check if `comps` is at least length of 4. Log at least debug if not, and include what was the value.

isbm commented on salt/modules/shadow.py:
```diff
@@ -306,81 +392,145 @@ def set_password(name, password, use_usermod=False):
     '''
-    return sorted([user.sp_nam for user in spwd.getspall()])
+    if root is not None:
+        _fun = functools.partial(_getspall, root=root)
```
I think, the same GvR's tweet still applies here. :wink:

isbm commented on salt/modules/shadow.py:
```diff
@@ -306,81 +392,145 @@ def set_password(name, password, use_usermod=False):
+                   for user in _fun()])
+
+
+def _getspnam(name, root=None):
```
This looks somewhat very similar to what is used in other module. You could join all that and make utility function instead, removing lots of code duplication from both places.

aplanas replied:
Ah indeed I was tempted to do this. Actually this patch removes a lot of old duplicated code. But in this case the interpretation of the columns in different for user, groups and shadow. The only generic part is open the file, read each line, and split the line with ':'

I will think what I can factor out.

isbm commented on salt/modules/useradd.py:
```diff
@@ -306,98 +381,132 @@ def getent(refresh=False):
     ret = []
-    for data in pwd.getpwall():
+    if root is not None and __grains__['kernel'] != 'AIX':
+        _fun = functools.partial(_getpwall, root=root)
```
:-)

isbm commented on salt/modules/useradd.py:
```diff
@@ -306,98 +381,132 @@ def getent(refresh=False):
+
+    cmd = ['usermod']
+
+    if root is not None and __grains__['kernel'] != 'AIX':
```
You mean, this is fine on Windows, OSX and Solaris?

aplanas replied:
nop, see `__virtual__`

isbm commented on salt/modules/useradd.py:
```diff
@@ -546,25 +703,43 @@ def chloginclass(name, loginclass, root=None):
+    # Please, note that if the non-root /etc/passwd file is long the
+    # iteration can be slow.
+    if root is not None and __grains__['kernel'] != 'AIX':
+        _fun = functools.partial(_getpwnam, root=root)
```
And again...

isbm commented on salt/modules/useradd.py:
```diff
@@ -654,43 +838,79 @@ def list_groups(name):
     '''
-    return sorted([user.pw_name for user in pwd.getpwall()])
+    if root is not None and __grains__['kernel'] != 'AIX':
+        _fun = functools.partial(_getpwall, root=root)
```
Bump.

isbm reviewed (changes requested):
@aplanas I see also lots of code duplication and repeating yourself. Could you please make sure code is not duplicated?

aplanas commented:
Any strong opinion to not merge this? There is one complain that can be fair: the new grp, pwd and spwd alternatives can be sharing some code (share the same structure of opening a file and reading the content line by line).

I do not want to merge them in the same code under the module utils, as they are very local. But I can revisit this opinion later if I see that this can be used outside their domain.

Also I removed a lot of [duplicated code](https://github.com/saltstack/salt/pull/50175/commits/1672f06d7e584c90ebed6bb24a9011f2a1a94470#diff-6ff7900860115f0b9224a20aca971b98L344) unrelated with my goal of introducing a new root parameter.

I am also not changing the semantic of the operations, validated with the old tests. This is something that we can revisit later, as makes sense to unify some fails (that sometimes are exceptions and other times return False)

aplanas commented:
I am still interested in this PR. Is there anything more that I can do?

aplanas commented:
* Rebased the code
* In the partial call I remove the ugly _fun name and make explicit the call to the correct function name in pwd, spwd and grp. I hope that this clarify why I currying over the real or the fake function (so I do not break the std interface)
* Refactor a bit _set_attrib to remove some extra returns. Still return early, as I am not able to find a way to avoid two nested indentations. But clarify the situation with a comment.

isbm reviewed (approved):
@cachedout I have no more issues with this PR. Unless someone else does, we're fine here.

aplanas commented:
Thanks for the reviews and the hard work here!
